packages/server/src/runtime/builtin/extract.rs

```diff
@@ -122,51 +125,72 @@ where
-					let file = tg::File::with_contents(blob);
-					let artifact = tg::Artifact::File(file);
-					Ok((path, artifact))
-				},
-				async_tar::EntryType::Symlink => {
-					let target = header
-						.link_name()
-						.map_err(|source| tg::error!(!source, "failed to read the symlink target"))?
-						.ok_or_else(|| tg::error!("no symlink target stored in the archive"))?;
-					let symlink = tg::Symlink::with_target(target.as_ref().into());
-					let artifact = tg::Artifact::Symlink(symlink);
-					Ok((path, artifact))
-				},
-				async_tar::EntryType::Directory => {
-					let directory = tg::Directory::with_entries([].into());
-					let artifact = tg::Artifact::Directory(directory);
-					Ok((path, artifact))
-				},
-				_ => Err(tg::error!("unsupported entry type")),
-			}
-		})
-		.try_collect()
-		.await?;
+		.map_err(|source| tg::error!(!source, "failed to get the entries from the archive"))?;
+
+	// Process the entries.
+	while let Some(entry) = entries.next().await {
+		let entry = entry
+			.map_err(|source| tg::error!(!source, "failed to get the entry from the archive"))?;
+		let header = entry.header();
+		let path = PathBuf::from(
+			entry
+				.path()
+				.map_err(|source| tg::error!(!source, "failed to get the entry path"))?
+				.as_ref(),
+		);
+		match header.entry_type() {
+			async_tar::EntryType::Regular | async_tar::EntryType::Other(_) => {
+				let blob = tg::Blob::with_reader(server, entry.compat()).await?;
+				let file = tg::File::with_contents(blob);
+				let artifact = tg::Artifact::File(file);
+				processed_entries.push((path, artifact));
+			},
+			async_tar::EntryType::Symlink => {
+				let target = header
+					.link_name()
+					.map_err(|source| tg::error!(!source, "failed to read the symlink target"))?
+					.ok_or_else(|| tg::error!("no symlink target stored in the archive"))?;
+				let symlink = tg::Symlink::with_target(target.as_ref().into());
+				let artifact = tg::Artifact::Symlink(symlink);
+				processed_entries.push((path, artifact));
+			},
+			async_tar::EntryType::Directory => {
+				let directory = tg::Directory::with_entries([].into());
+				let artifact = tg::Artifact::Directory(directory);
+				processed_entries.push((path, artifact));
+			},
+			async_tar::EntryType::Link => {
+				let target = header
+					.link_name()
+					.map_err(|source| {
+						tg::error!(!source, "failed to read the hard link target path")
+					})?
+					.ok_or_else(|| tg::error!("no hard link target path stored in the archive"))?;
+				let target: &Path = target.as_ref().into();
+				if let Some((_, target_artifact)) =
+					processed_entries.iter().find(|(path, _)| path == target)
+				{
+					processed_entries.push((path, target_artifact.clone()));
+				} else {
+					return Err(tg::error!("could not find hard link target in the archive"));
+				}
+			},
+			async_tar::EntryType::XGlobalHeader => {
+				continue;
+			},
+			_ => return Err(tg::error!("unsupported entry type")),
```

Comment:
small nit, but it might be useful to add the type to the error?

Reply:
Removed the error.